File: native/BackgroundRecorderPlugin.java
package com.pitchrec.backgroundrecorder;

import android.content.Intent;
import android.content.Context;
import android.net.Uri;
import android.os.Build;
import android.os.PowerManager;
import android.provider.Settings;

import com.getcapacitor.JSObject;
import com.getcapacitor.Plugin;
import com.getcapacitor.PluginCall;
import com.getcapacitor.PluginMethod;
import com.getcapacitor.annotation.CapacitorPlugin;
import com.getcapacitor.annotation.Permission;
import com.getcapacitor.annotation.PermissionCallback;
import com.getcapacitor.PermissionState;

@CapacitorPlugin(
    name = "BackgroundRecorder",
    permissions = {
        @Permission(strings = { android.Manifest.permission.RECORD_AUDIO }, alias = "microphone")
    }
)
public class BackgroundRecorderPlugin extends Plugin {

    static PluginCall pendingStopCall;

    @PluginMethod
    public void hasAudioRecordingPermission(PluginCall call) {
        boolean granted = getPermissionState("microphone") == PermissionState.GRANTED;
        JSObject ret = new JSObject();
        ret.put("value", granted);
        call.resolve(ret);
    }

    @PluginMethod
    public void requestAudioRecordingPermission(PluginCall call) {
        if (getPermissionState("microphone") == PermissionState.GRANTED) {
            JSObject ret = new JSObject();
            ret.put("value", true);
            call.resolve(ret);
            return;
        }
        requestPermissionForAlias("microphone", call, "permissionCallback");
    }

    @PermissionCallback
    private void permissionCallback(PluginCall call) {
        boolean granted = getPermissionState("microphone") == PermissionState.GRANTED;
        JSObject ret = new JSObject();
        ret.put("value", granted);
        call.resolve(ret);
    }

    @PluginMethod
    public void startRecording(PluginCall call) {
        if (getPermissionState("microphone") != PermissionState.GRANTED) {
            call.reject("MISSING_PERMISSION");
            return;
        }
        try {
            Intent intent = new Intent(getContext(), BackgroundRecorderService.class);
            intent.setAction(BackgroundRecorderService.ACTION_START);
            if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.O) {
                getContext().startForegroundService(intent);
            } else {
                getContext().startService(intent);
            }
            JSObject ret = new JSObject();
            ret.put("value", true);
            call.resolve(ret);
        } catch (Exception e) {
            call.reject("FAILED_TO_RECORD", e.getMessage());
        }
    }

    @PluginMethod
    public void pauseRecording(PluginCall call) {
        sendAction(BackgroundRecorderService.ACTION_PAUSE);
        JSObject ret = new JSObject();
        ret.put("value", true);
        call.resolve(ret);
    }

    @PluginMethod
    public void resumeRecording(PluginCall call) {
        sendAction(BackgroundRecorderService.ACTION_RESUME);
        JSObject ret = new JSObject();
        ret.put("value", true);
        call.resolve(ret);
    }

    @PluginMethod
    public void stopRecording(PluginCall call) {
        pendingStopCall = call;
        Intent intent = new Intent(getContext(), BackgroundRecorderService.class);
        intent.setAction(BackgroundRecorderService.ACTION_STOP);
        getContext().startService(intent);
    }

    @PluginMethod
    public void getCurrentStatus(PluginCall call) {
        JSObject ret = new JSObject();
        ret.put("status", BackgroundRecorderService.currentStatus);
        call.resolve(ret);
    }

    @PluginMethod
    public void isIgnoringBatteryOptimizations(PluginCall call) {
        PowerManager pm = (PowerManager) getContext().getSystemService(Context.POWER_SERVICE);
        boolean ignoring = true;
        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.M) {
            ignoring = pm.isIgnoringBatteryOptimizations(getContext().getPackageName());
        }
        JSObject ret = new JSObject();
        ret.put("value", ignoring);
        call.resolve(ret);
    }

    @PluginMethod
    public void requestIgnoreBatteryOptimizations(PluginCall call) {
        if (Build.VERSION.SDK_INT < Build.VERSION_CODES.M) {
            JSObject ret = new JSObject();
            ret.put("value", true);
            call.resolve(ret);
            return;
        }
        PowerManager pm = (PowerManager) getContext().getSystemService(Context.POWER_SERVICE);
        if (pm.isIgnoringBatteryOptimizations(getContext().getPackageName())) {
            JSObject ret = new JSObject();
            ret.put("value", true);
            call.resolve(ret);
            return;
        }
        try {
            Intent intent = new Intent(Settings.ACTION_REQUEST_IGNORE_BATTERY_OPTIMIZATIONS);
            intent.setData(Uri.parse("package:" + getContext().getPackageName()));
            getActivity().startActivity(intent);
            JSObject ret = new JSObject();
            ret.put("value", true);
            call.resolve(ret);
        } catch (Exception e) {
            call.reject("FAILED_TO_REQUEST", e.getMessage());
        }
    }

    private void sendAction(String action) {
        Intent intent = new Intent(getContext(), BackgroundRecorderService.class);
        intent.setAction(action);
        getContext().startService(intent);
    }

    static void resolveStop(String base64, long durationMs, String mimeType) {
        JSObject ret = new JSObject();
        ret.put("recordDataBase64", base64);
        ret.put("msDuration", durationMs);
        ret.put("mimeType", mimeType);
        if (pendingStopCall != null) {
            pendingStopCall.resolve(ret);
            pendingStopCall = null;
        }
    }

    static void rejectStop(String code, String message) {
        if (pendingStopCall != null) {
            pendingStopCall.reject(code, message);
            pendingStopCall = null;
        }
    }
}

File: native/BackgroundRecorderPlugin.kt
package com.pitchrec.backgroundrecorder

import android.content.Intent
import android.os.Build
import com.getcapacitor.JSObject
import com.getcapacitor.Plugin
import com.getcapacitor.PluginCall
import com.getcapacitor.PluginMethod
import com.getcapacitor.annotation.CapacitorPlugin
import com.getcapacitor.annotation.Permission
import com.getcapacitor.annotation.PermissionCallback

@CapacitorPlugin(
    name = "BackgroundRecorder",
    permissions = [
        Permission(strings = [android.Manifest.permission.RECORD_AUDIO], alias = "microphone")
    ]
)
class BackgroundRecorderPlugin : Plugin() {

    @PluginMethod
    fun hasAudioRecordingPermission(call: PluginCall) {
        val granted = getPermissionState("microphone") == com.getcapacitor.PermissionState.GRANTED
        val ret = JSObject()
        ret.put("value", granted)
        call.resolve(ret)
    }

    @PluginMethod
    fun requestAudioRecordingPermission(call: PluginCall) {
        if (getPermissionState("microphone") == com.getcapacitor.PermissionState.GRANTED) {
            val ret = JSObject()
            ret.put("value", true)
            call.resolve(ret)
            return
        }
        requestPermissionForAlias("microphone", call, "permissionCallback")
    }

    @PermissionCallback
    private fun permissionCallback(call: PluginCall) {
        val granted = getPermissionState("microphone") == com.getcapacitor.PermissionState.GRANTED
        val ret = JSObject()
        ret.put("value", granted)
        call.resolve(ret)
    }

    @PluginMethod
    fun startRecording(call: PluginCall) {
        if (getPermissionState("microphone") != com.getcapacitor.PermissionState.GRANTED) {
            call.reject("MISSING_PERMISSION")
            return
        }
        try {
            val intent = Intent(context, BackgroundRecorderService::class.java)
            intent.action = BackgroundRecorderService.ACTION_START
            if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.O) {
                context.startForegroundService(intent)
            } else {
                context.startService(intent)
            }
            val ret = JSObject()
            ret.put("value", true)
            call.resolve(ret)
        } catch (e: Exception) {
            call.reject("FAILED_TO_RECORD", e.message)
        }
    }

    @PluginMethod
    fun pauseRecording(call: PluginCall) {
        sendAction(BackgroundRecorderService.ACTION_PAUSE)
        val ret = JSObject()
        ret.put("value", true)
        call.resolve(ret)
    }

    @PluginMethod
    fun resumeRecording(call: PluginCall) {
        sendAction(BackgroundRecorderService.ACTION_RESUME)
        val ret = JSObject()
        ret.put("value", true)
        call.resolve(ret)
    }

    @PluginMethod
    fun stopRecording(call: PluginCall) {
        pendingStopCall = call
        val intent = Intent(context, BackgroundRecorderService::class.java)
        intent.action = BackgroundRecorderService.ACTION_STOP
        context.startService(intent)
    }

    @PluginMethod
    fun getCurrentStatus(call: PluginCall) {
        val ret = JSObject()
        ret.put("status", BackgroundRecorderService.currentStatus)
        call.resolve(ret)
    }

    private fun sendAction(action: String) {
        val intent = Intent(context, BackgroundRecorderService::class.java)
        intent.action = action
        context.startService(intent)
    }

    companion object {
        var pendingStopCall: PluginCall? = null

        fun resolveStop(base64: String, durationMs: Long, mimeType: String) {
            val ret = JSObject()
            ret.put("recordDataBase64", base64)
            ret.put("msDuration", durationMs)
            ret.put("mimeType", mimeType)
            pendingStopCall?.resolve(ret)
            pendingStopCall = null
        }

        fun rejectStop(code: String, message: String?) {
            pendingStopCall?.reject(code, message)
            pendingStopCall = null
        }
    }
}
